src/codecs/json.rs

```diff
@@ -0,0 +1,176 @@
+                }
+
+                let target_field = options
+                    .target_field
+                    .as_ref()
+                    .map(AsRef::as_ref)
+                    .unwrap_or_else(|| log_schema().message_key());
+
+                let contains_target = log.contains(&target_field);
+                if !contains_target || options.overwrite_target {
+                    log.insert(&target_field, serialized);
+                }
+
+                output.push(event)
+            }
+        }
+
+        Ok(CodecTransform {
+            transform: Transform::function(SerializeJsonTransform(self.clone())),
+            input_type: DataType::Log,
+        })
+    }
+}
+
+#[typetag::serde(name = "json")]
+impl Codec for JsonCodec {
+    fn name(&self) -> &'static str {
+        "json"
+    }
+
+    fn build(&self, hint: CodecHint) -> crate::Result<CodecTransform> {
+        match (self.mode, hint) {
+            (Some(Mode::Decode), _) | (None, CodecHint::Decoder) => self.build_decoder(),
+            (Some(Mode::Encode), _) | (None, CodecHint::Encoder) => self.build_encoder(),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum Mode {
+    Encode,
+    Decode,
+}
+
+inventory::submit! {
+    Box::new(JsonCodec::default()) as Box<dyn Codec>
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use futures::stream::{self, StreamExt};
+    use shared::btreemap;
+
+    #[tokio::test]
+    async fn default_decoder() {
+        let codec = JsonCodec::default();
+        let transform = codec.build(CodecHint::Decoder).unwrap().transform;
+        let mut event = Event::new_empty_log();
+        let log = event.as_mut_log();
+        log.insert(log_schema().message_key(), r#"{"foo":"bar"}"#);
+        let input = stream::once(async { event });
+        let mut output = transform.transform(input);
+        let transformed = output.next().await.unwrap();
+
+        assert_eq!(
+            transformed,
+            btreemap! {
+                "foo" => "bar"
+            }
+            .into(),
+        );
+        assert_eq!(output.next().await, None);
+    }
+
+    #[tokio::test]
+    async fn default_encoder() {
+        let codec = JsonCodec::default();
+        let transform = codec.build(CodecHint::Encoder).unwrap().transform;
+        let mut event = Event::new_empty_log();
+        let log = event.as_mut_log();
+        log.insert(
+            log_schema().message_key(),
+            btreemap! {
+                "foo" => "bar"
+            },
+        );
+        let input = stream::once(async { event });
+        let mut output = transform.transform(input);
+        let transformed = output.next().await.unwrap();
+
+        assert_eq!(
+            transformed
+                .as_log()
+                .get(log_schema().message_key())
+                .unwrap()
+                .clone(),
+            r#"{"message":{"foo":"bar"}}"#.to_owned().into()
```

Comment:
I'm a bit unsure about exposing the "message" field when not specifying the source field. However, there doesn't seem to be an unambiguous way to differentiate between an object on the root that incidentally has a message key, or the "main" message key from the log schema.